### PR TITLE
Feature/edit mode

### PR DESCRIPTION
**Pull request by VSmain**

### CODE REVIEW

File: src/routes/api/edited.js [hunk 1]
```diff
@@ -42,11 +46,21 @@ router.delete("/delete", (req, res) => {
...
+      let fn = LambdaBuilder.buildContext(req.body.map, context);
+      let result = fn.call(context, req.body.p);
+      return resolve(result);
+    }
+  )
     .then(result => res.send(result))
-    .catch(err => res.status(400).send({error: err.message}))
+    .catch(err => res.status(400).send({error: err.message}));
+
+  // editedPlaceService.buildFunctions(req.body)
```
Comment (aartemenko89) [inline src/routes/api/edited.js]: remove

File: src/routes/api/edited.js [hunk 2]
```diff
@@ -42,11 +46,21 @@ router.delete("/delete", (req, res) => {
   editedPlaceService.deleteByIds(req.body).then(result => res.send(result)).catch(err => res.send(err))
 });
 
-router.post("/map",(req,res)=>{
-  editedPlaceService.buildFunctions(req.body)
-    .then(r=>editedPlaceService.evaluateLogic(r))
+router.post("/map/test", (req, res) => {
+  return new Promise((resolve, reject) => {
+      let context = _.clone(LambdaBuilder.getFunctionContext());
```
Comment (aartemenko89) [inline src/routes/api/edited.js]: const

File: src/services/lambdaBuilder.ts [hunk 3]
```diff
@@ -2,93 +2,196 @@ import Constants from '../utils/constants';
...
+export default class LambdaBuilder {
 
-  public static buildMapFunction(lambdas: any, sequence: string[]): Function {
-    let functions = [];
-    sequence.forEach(k => {
-      if (lambdas.get(k) != undefined)
-        functions.push(lambdas.get(k))
-    });
-    return (p, _) => {
+  private static NOT_DEFINED: string = " not defined";
```
Comment (aartemenko89) [inline src/services/lambdaBuilder.ts]: use single quat...

File: src/services/lambdaBuilder.ts [hunk 4]
```diff
@@ -2,93 +2,196 @@ import Constants from '../utils/constants';
...
-        funcMap.set(key, treeFuncMap.get(key))
-      });
-      return funcMap;
-    } catch (e) {
-      throw e;
+  private static buildFunctionBody(definition: any): Function {
+    let bodyVsParams = definition[0];
+    let params: string[] = ["p"];
+    let body: string = "";
+    if (bodyVsParams.length == 1) {
```
Comment (aartemenko89) [inline src/services/lambdaBuilder.ts]: === 1

Comment (aartemenko89) [reply src/services/lambdaBuilder.ts]: > 0

File: src/services/lambdaBuilder.ts [hunk 5]
```diff
@@ -2,93 +2,196 @@ import Constants from '../utils/constants';
...
+      body = this.tryAddReturnKeyword(bodyVsParams[0]);
     }
+    if (bodyVsParams.length > 1) {
+      let i = 0;
+      params = [];
+      for (; i < bodyVsParams.length - 1; i++) {
```
Comment (aartemenko89) [inline src/services/lambdaBuilder.ts]: clone  - 1

File: src/services/lambdaBuilder.ts [hunk 6]
```diff
@@ -2,93 +2,196 @@ import Constants from '../utils/constants';
...
+    return _.isArray(definition) && (_.isString(definition[0]) || _.isObject(definition[0]))
+  }
+
+  private static isAutoMergeForDecisionTree(definition: any): boolean {
+    return _.isArray(definition) && definition.length == 1 && _.isObject(definition[0]);
+  }
+
+  private static useAutoMerge(definition: any): boolean {
+    if (this.isParallelFunctionsDefinition(definition))
+      return !definition.some(v => v == null);
```
Comment (aartemenko89) [inline src/services/lambdaBuilder.ts]: lodash

File: src/services/lambdaBuilder.ts [hunk 7]
```diff
@@ -2,93 +2,196 @@ import Constants from '../utils/constants';
...
+  private static isParallelFunctionsDefinition(definition: any): boolean {
+    return _.isArray(definition) && (_.isString(definition[0]) || _.isObject(definition[0]))
+  }
+
+  private static isAutoMergeForDecisionTree(definition: any): boolean {
+    return _.isArray(definition) && definition.length == 1 && _.isObject(definition[0]);
+  }
+
+  private static useAutoMerge(definition: any): boolean {
+    if (this.isParallelFunctionsDefinition(definition))
```
Comment (aartemenko89) [inline src/services/lambdaBuilder.ts]: this.isParallelFunctionsDefinition(definition) && !definition.some(v => v == null)

File: src/services/br-engine/lambdaBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  private static useAutoMerge(definition: any): boolean {
+    return this.isParallelFunctionsDefinition(definition) && !definition.some(v => v === null)
+  }
+
+  private static createError(alias: string, body: any, err: string): Error {
+    return new Error(`alias ${alias} \n ${JSON.stringify(body)}  \n ${err}`)
+  }
+
+  private static collectResults(functions: Function[], p: any, context: any): any[] {
+    let results = functions.map(fn => fn.call(context, _.cloneDeep(p))).filter(v => v != null);
```
Comment (aartemenko89) [inline src/services/br-engine/lambdaBuilder.ts]: https://lodash.com/docs/4.17.10#isNull

File: src/services/br-engine/lambdaBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  private static joinAliasChain(aliasChain: string[], context: any): Function {
+    let functions = aliasChain.map(alias => context[alias]);
+
+    return p => {
+      const params = _.cloneDeep(p);
+
+      let r;
+      if (params === null)
```
Comment (aartemenko89) [inline src/services/br-engine/lambdaBuilder.ts]: https://lodash.com/docs/4.17.10#isNull

File: src/services/br-engine/lambdaBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        return null;
+
+      for (let f of functions) {
+        r = f.call(context, params);
+        if (r === false || r === null) {
```
Comment (aartemenko89) [inline src/services/br-engine/lambdaBuilder.ts]: mb !r

File: src/utils/constants.ts [hunk 11]
```diff
@@ -29,7 +32,7 @@ export default class Constants {
 
   public static readonly AWS_PLACE_BUCKET: string = process.env.AWS_PLACE_BUCKET;
   public static readonly AWS_PLACE_EXPORT_KEY: string = 'unikoom_us_places.exported.csv';
-  public static readonly AWS_CLUSTER_KEY: string = 'unikoom_us_places.cluster.csv';
+  public static readonly AWS_CLUSTER_KEY: string = 'at_us_places.cluster.csv';
```
Comment (TsapAndrii) [inline src/utils/constants.ts]: at_us_places -> keep that only local

**Pull request merged**